File: SearchA2DMatrix.java
//Approach : In each row, we check that if the target is <= last element only then we do a bonary search on that row otherwise we go to the next row
//Time - O(m + log n) where m is the number of rows and n is the number of columns
//Space - O(1)
public class SearchA2DMatrix {
    public boolean searchMatrix(int[][] nums, int target) {
        if(nums==null || nums.length==0) return false;

        for(int i=0; i<nums.length; i++){
            if(target > nums[i][nums[i].length-1]) continue;
            int left = 0, right = nums[0].length-1;
            while(left<=right){
                int mid = left + ((right-left)/2);
                if(nums[i][mid] == target){
                    return true;
                } else if(nums[i][mid] < target){
                    left = mid+1;
                } else{
                    right = mid-1;
                }
            }
        }

        return false;
    }
}

File: Search_a _2D_Matrix.py
# Time complexity: O(logm + logn)
# Space complexity: O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False 
        m = len(matrix) # number of rows
        n = len(matrix[0]) # number of columns

        # binary search on row
        top = 0
        bottom = m-1
        while top <= bottom:
            mid = (top+bottom)//2
            if matrix[mid][0] <= target <= matrix[mid][-1]:
                left = 0
                right = n-1
                while left <= right:
                    mid1 = (left+right)//2
                    if matrix[mid][mid1] == target:
                        return True
                    elif matrix[mid][mid1] < target:
                        left = mid1+1
                    else:
                        right = mid1-1
                return False

            elif target < matrix[mid][0]:
                bottom = mid-1
            else:
                top = mid + 1
        return False
        